servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherConcatWithCompletable.java

```diff
@@ -31,20 +31,154 @@
  */
 final class PublisherConcatWithCompletable<T> extends AbstractAsynchronousPublisherOperator<T, T> {
     private final Completable next;
+    private final boolean propagateCancel;
 
-    PublisherConcatWithCompletable(Publisher<T> original, Completable next) {
+    PublisherConcatWithCompletable(Publisher<T> original, Completable next, boolean propagateCancel) {
         super(original);
         this.next = requireNonNull(next);
+        this.propagateCancel = propagateCancel;
     }
 
     @Override
     public Subscriber<? super T> apply(Subscriber<? super T> subscriber) {
-        return new ConcatSubscriber<>(subscriber, next);
+        return propagateCancel ?
+                new ConcatSubscriberCancel<>(subscriber, next) : new ConcatSubscriber<>(subscriber, next);
+    }
+
+    private static final class ConcatSubscriberCancel<T>
+            implements CompletableSource.Subscriber, PublisherSource.Subscriber<T>, Subscription {
+        private static final Cancellable CANCELLED = () -> { };
+        @SuppressWarnings("rawtypes")
+        private static final AtomicReferenceFieldUpdater<ConcatSubscriberCancel, Cancellable> cancellableUpdater =
+                newUpdater(ConcatSubscriberCancel.class, Cancellable.class, "cancellable");
+        private final Subscriber<? super T> target;
+        private final Completable next;
+        private volatile Cancellable cancellable = IGNORE_CANCEL;
+
+        ConcatSubscriberCancel(Subscriber<? super T> target, Completable next) {
+            this.target = target;
+            this.next = next;
+        }
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            cancellable = new FirstSubscription(s);
```

Comment:
Can we use `instanceof Subscription` instead of wrapping, like `ConcatSubscriber` does? If not, please leave a comment in code to clarify why we need it.

Reply:
We need to differentiate between the first `onSubscribe(Subscription)` and the second `onSubscribe(Cancellable)`. someone could pass a `Cancellable` that is also a `Subscription` to the second method so we need some additional state.

Reply:
added comments